File: apps/web/lib/api/entities.ts
import { API_URL } from "../utils";

export interface FieldMetadata {
  key: string;
  label: string;
  data_type: string;
  description?: string;
  options?: string[];
  is_custom: boolean;
  validation?: Record<string, any>;
}

export interface EntityMetadata {
  key: string;
  label: string;
  fields: FieldMetadata[];
}

export interface OperatorDefinition {
  value: string;
  label: string;
  field_types: string[];
}

export interface EntitiesMetadataResponse {
  entities: EntityMetadata[];
  operators: OperatorDefinition[];
}

export async function fetchEntitiesMetadata(): Promise<EntitiesMetadataResponse> {
  const response = await fetch(`${API_URL}/entities/metadata`);
  if (!response.ok) {
    throw new Error("Failed to fetch entities metadata");
  }
  return response.json();
}

/**
 * Fetch entity data for tooltip display
 *
 * Used by EntityTooltip component to lazy-load entity details on hover.
 *
 * @param entityType - Type of entity (cpu, gpu, ram-spec, storage-profile)
 * @param entityId - ID of the entity to fetch
 * @returns Promise resolving to entity data
 */
export async function fetchEntityData(
  entityType: string,
  entityId: number
): Promise<any> {
  const endpoints: Record<string, string> = {
    cpu: `/v1/catalog/cpus/${entityId}`,
    gpu: `/v1/catalog/gpus/${entityId}`,
    "ram-spec": `/v1/catalog/ram-specs/${entityId}`,
    "storage-profile": `/v1/catalog/storage-profiles/${entityId}`,
  };

  const endpoint = endpoints[entityType];
  if (!endpoint) {
    throw new Error(`Unknown entity type: ${entityType}`);
  }

  const url = `${API_URL}${endpoint}`;

  try {
    const response = await fetch(url);

    if (!response.ok) {
      throw new Error(
        `Failed to fetch ${entityType}: ${response.status} ${response.statusText}`
      );
    }

    return response.json();
  } catch (error) {
    console.error(`fetchEntityData error for ${entityType} ${entityId}:`, error);
    throw error;
  }
}

File: apps/api/dealbrain_api/api/entities.py
"""Endpoints for entity and field metadata."""

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from ..db import session_dependency
from ..services.field_metadata import FieldMetadataService


router = APIRouter(prefix="/entities", tags=["entities"])


@router.get("/metadata")
async def get_entities_metadata(
    db: AsyncSession = Depends(session_dependency),
) -> dict:
    """Get metadata for all entities and fields for rule builder."""
    service = FieldMetadataService()
    entities = await service.get_entities_metadata(db)

    return {
        "entities": [
            {
                "key": entity.key,
                "label": entity.label,
                "fields": [
                    {
                        "key": field.key,
                        "label": field.label,
                        "data_type": field.data_type,
                        "description": field.description,
                        "options": field.options,
                        "is_custom": field.is_custom,
                        "validation": field.validation,
                    }
                    for field in entity.fields
                ],
            }
            for entity in entities
        ],
        "operators": [
            {
                "value": op.value,
                "label": op.label,
                "field_types": op.field_types,
            }
            for op in FieldMetadataService.OPERATORS
        ],
    }
